[inductor] post_grad batched linear fusion

xuzhao9 opened the pull request:
Summary: Fusing independent nn.Linear() functions with aten.bmm and aten.cat.

Test Plan:
Without the BMM fusion:
```
buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 0
```
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072536_6535183793.json.gz&bucket=pyper_traces

100 aten::mm operators

With the BMM fusion:
```
buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1
```

20 aten::bmm operators

https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072157_6535183793.json.gz&bucket=pyper_traces

Passes accuracy test:
```
$ buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --accuracy
Running eval method from test_module on cuda in dynamo inductor mode with input batch size 4 and precision tf32.
Accuracy:                            pass
```
Looks like the bmm and input cat has been fused successfully.

Checking the triton codegen:

```
TORCH_LOGS=+dynamo,+aot,+inductor buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --dump_triton 1
```

Triton code dump: https://www.internalfb.com/intern/everpaste/?handle=GHp1ABaqYuTjYCUBALiTWmteaI1PbsIXAAAB




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

xuzhao9 commented:
@yanboliang Can I get a review on this PR?

yanboliang commented on torch/_inductor/config.py:
```diff
@@ -79,6 +79,9 @@
 
+# enable post-grad pattern match with batch fusion (using torch op)
+post_grad_batch_fusion = False
+
```
I don't think we have to add a separate flag for this, we already have ```batch_fusion``` at L80.

xuzhao9 replied:
I thought `batch_fusion` at L80 is pre_grad? I am thinking it would still be useful to make them separate config flags for debugging purpose.

yanboliang commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -66,6 +66,89 @@ class BatchFusion(GroupBatchFusionBase):
     pass
 
 
+class BatchLinearPostGradFusion(GroupBatchFusionBase):
```
Probably we have to update the original ```BatchLinearFusion``` to ```PreGradBatchLinearFusion``` and name this as ```PostGradBatchLinearFusion```.

yanboliang commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -653,6 +739,9 @@ def group_batch_fusion_post_grad_passes(graph: torch.fx.Graph):
 
+    if config.post_grad_batch_fusion:
+        fusions += [BatchLinearPostGradFusion()]
+
```
I think we need a mechanism to only enable one batch linear fusion, either pre or post grad.

yanboliang replied:
Or probably we have a config can be used to specify a list of fusions and the order.

yanboliang commented on torch/_inductor/fx_passes/group_batch_fusion.py:
```diff
@@ -722,14 +806,10 @@ def group_batch_fusion_passes(graph: torch.fx.Graph, pre_grad=True):
+    fusion_options = (
+        config.pre_grad_fusion_options if pre_grad else config.post_grad_fusion_options
+    )
+    fusions = generate_fusion_from_config(fusion_options, pre_grad=pre_grad)
```
We need to check ```has_fbgemm```, as we can't run group fusion if fbgemm is not available. I'd like to add a ```need_fbgemm``` attribute to ```GroupBatchFusionBase``` and enable the fusion pass if the check passed.

xuzhao9 replied:
Right, I will fix this on top of https://github.com/pytorch/pytorch/pull/114778

xuzhao9 replied:
Hi @yanboliang , I have addressed your comment, can you help take another look?

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge

xuzhao9 committed:
[inductor] post_grad batched linear fusion (#112504)

Summary:

Fusing independent nn.Linear() functions with aten.bmm and aten.cat.

Test Plan:
Without the BMM fusion:
```
buck2 run mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 0
```
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072536_6535183793.json.gz&bucket=pyper_traces

100 aten::mm operators

With the BMM fusion:
```
buck2 run mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1
```

20 aten::bmm operators

https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072157_6535183793.json.gz&bucket=pyper_traces

Passes accuracy test:
```
$ buck2 run mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --accuracy
Running eval method from test_module on cuda in dynamo inductor mode with input batch size 4 and precision tf32.
Accuracy:                            pass
```
Looks like the bmm and input cat has been fused successfully.

Checking the triton codegen:

```
TORCH_LOGS=+dynamo,+aot,+inductor buck2 run mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --dump_triton 1
```

Triton code dump: https://www.internalfb.com/intern/everpaste/?handle=GHp1ABaqYuTjYCUBALiTWmteaI1PbsIXAAAB

Reviewed By: yanboliang, jackiexu1992

Differential Revision: D46910718

xuzhao9 commented:
@pytorchbot merge

xuzhao9 commented:
@pytorchbot merge -f "Skip failed lintrunner job timeout. Need to land this to unblock another prod fix"
